README.md
# google4r/checkout [![Build Status](https://secure.travis-ci.org/nbudin/google4r-checkout.png)](http://travis-ci.org/nbudin/google4r-checkout)

google4r/checkout is a library to access the Google Checkout API.

It currently supports version 2.3 of the API and subscription beta API.

### License

google4r itself is distributed under an MIT style license.

However, the library includes the [cacert.pem](http://curl.haxx.se/ca/cacert.pem) file from the Mozilla project. This file is distributed under the [MPL](http://www.mozilla.org/MPL/).

### Installing

Gems are hosted on rubygems.org (aka Gemcutter), so on reasonably recent versions of Rubygems, you should be able to install just like this:

    gem install google4r-checkout

Or, go to [our page on rubygems.org](http://rubygems.org/gems/google4r-checkout).

### Issue Tracking and Wiki

Our issue tracker and wiki can be found [on Google Code](http://code.google.com/p/google-checkout-ruby-sample-code/).  The best way to let us know about bugs or feature requests is to report an issue there.

### Documentation

We've got RDoc documentation for the google4r-checkout library [generated on rdoc.info](http://rdoc.info/projects/nbudin/google4r-checkout).

You can find more information on the Google Checkout API [here](http://code.google.com/apis/checkout/developer/index.html). Note that the API documentation assumes an understanding of the Google Checkout XML API.

## Google Checkout Tests

You have to place a file called 'frontend_configuration.rb' in the directory'test' with the configuration for the Google4R::Checkout::Frontend class to use for running the tests.

The file should contain content similar to:

```ruby
# Uncomment the following line if you are using Google Checkout in Great Britain
# and adjust it if you want to test google4r-checkout against any other (future)
# Google Checkout service.
    
# Money.default_currency = 'GBP'

# The test configuration for the Google4R::Checkout::Frontend class.
FRONTEND_CONFIGURATION = 
  { 
    :merchant_id => '<your merchant id>', 
    :merchant_key => '<your merchant key>',
    :use_sandbox => true
  }
```

## Sending Commands to Google Checkout

To send commands to Google Checkout, use a Google4R::Checkout::Frontend object.  The Frontend class contains a variety of methods for easily generating several types of commands, including checkout, cancel_order, charge_and_ship_order, etc.

Here's an example:

```ruby
# Create the Frontend from our configuration
frontend = Google4R::Checkout::Frontend.new(
  :merchant_id => conf['merchant_id'],
  :merchant_key => conf['merchant_key'],
  :use_sandbox => conf['use_sandbox']
)
      
# Create a new checkout command (to place an order)
cmd = frontend.create_checkout_command
      
# Add an item to the command's shopping cart
cmd.shopping_cart.create_item do |item|
  item.name = "2-liter bottle of Diet Pepsi"
  item.quantity = 100
  item.unit_price = Money.new(1.99, "USD")
end
      
# Send the command to Google and capture the HTTP response
begin
  response = cmd.send_to_google_checkout
rescue Net::HTTPServerError => e
  # Sometimes Google Checkout is unavailable for internal reasons.
  # Because of this, it's a good idea to catch Net::HTTPServerError
  # and retry.
  
  response = cmd.send_to_google_checkout
end
      
# Redirect the user to Google Checkout to complete the transaction
redirect_to response.redirect_url
```

For more information, see the Frontend class's RDocs and the Google Checkout API documentation.

## Writing Responders

To receive notifications from Google Checkout, you'll need to write an action in your web application to respond to Google Checkout XML requests.  The basic strategy for this is to use a NotificationHandler object (which can be obtained through a Frontend object) to parse the HTTP request, then handle the resulting Notification object according to its class.  You should then generate a NotificationAcknowledgment and return it as the response.

It's a good idea to verify the HTTP authentication headers for incoming requests.  This will be a combination of your Merchant ID and Merchant Key.

Here is an example of how one might do this in Rails:

```ruby
class PaymentNotificationController < ApplicationController
  before_filter :verify_merchant_credentials, :only => [:google]
  
  def google
    frontend = Google4R::Checkout::Frontend.new(
      :merchant_id => conf['merchant_id'],
      :merchant_key => conf['merchant_key'],
      :use_sandbox => conf['use_sandbox']
    )
    handler = frontend.create_notification_handler
    
    begin
       notification = handler.handle(request.raw_post) # raw_post contains the XML
    rescue Google4R::Checkout::UnknownNotificationType
       # This can happen if Google adds new commands and Google4R has not been
       # upgraded yet. It is not fatal.
       logger.warn "Unknown notification type"
       return render :text => 'ignoring unknown notification type', :status => 200
    end
    
    case notification
    when Google4R::Checkout::NewOrderNotification then
      
      # handle a NewOrderNotification
      
    when Google4R::Checkout::OrderStateChangeNotification then
      
      # handle an OrderStateChangeNotification
      
    else
      return head :text => "I don't know how to handle a #{notification.class}", :status => 500
    end

    notification_acknowledgement = Google4R::Checkout::NotificationAcknowledgement.new(notification)
    render :xml => notification_acknowledgement.to_xml, :status => 200
  end

  private
  # make sure the request authentication headers use the right merchant_id and merchant_key
  def verify_merchant_credentials
    authenticate_or_request_with_http_basic("Google Checkout notification endpoint") do |merchant_id, merchant_key|
      (conf['merchant_id'].to_s == merchant_id.to_s) and (conf['merchant_key'].to_s == merchant_key.to_s)
    end
  end
end
```

## Dependencies

google4r-checkout makes extensive use of the money gem library, which is a required dependency.  The unit tests also use Mocha and Nokogiri, which should be automatically pulled down by Bundler.

google4r-checkout doesn't depend on any particular Ruby web framework, so it should work with any version of Rails, Sinatra, Camping, or even no web framework at all.


CHANGES
=google4r-checkout Changelog

== 1.2.0 (2013-04-09)

* Support for the Polling API, contributed by Chance Downs and Jacob Comer

== 1.1.6 (2012-08-01)

* Bugfixes when creating recurring subscriptions parsed from XML, contributed by James Martin

== 1.1.5 (2012-08-01)

* Raise a more informative exception for non-XML notifications, contributed by James Martin

== 1.1.4 (2012-07-31)

* Carrier shipping adjustment support, contributed by Ben Hutton

== 1.1.3 (2012-07-11)

* More complete Notification History API support, contributed by Ben Hutton

== 1.1.2 (2012-05-31)

* Added support for donation processing, contributed by Paul Schreiber

== 1.1.1 (2011-12-16)

* Bug fix for subscriptions not being processed, contributed by Larry Salibra
* Automated testing with Travis
* Officially support JRuby

== 1.1.0 (2011-06-29)

* Support for newer refund-notification XML format, contributed by Chris Parrish
* Update README to remove outdated Rails gem locking technique and add more examples
* Modernize gem build

== 1.1.beta5 (2010-09-21)

* Support for chargeback fees and chargeback gateway refund fees, contributed by George Palmer
* Test::Unit fixes for Ruby 1.9.2

== 1.1.beta4 (2010-07-09)

* All changes in this release contributed by George Palmer (thanks very much George!)
* Make the tax field optional as can be set at Google Account level
* Add a new command ChargeAndShipOrder (v2.5 of the API although we ended up going with 2.4 but it was fully tested and worked)
* Ability to get the refund amount available in RefundAmountNotification
* Ability to get gateway fee information

== 1.1.beta3 (2010-06-28)

* Add required test files to the gem build

== 1.1.beta2 (2010-06-23)

* Make gem compatible with Money 3.x (thanks George Palmer!)
* Use Bundler for running test suite

== 1.1.beta1 (2010-04-21)

* Full Ruby 1.9.x compatibility
* Major overhaul of automated test suite
* Support for merchant-calculated and allow-us-po-box tags
* Update to Money library version 2.3 (beware - this may break things if you expect the currency to be a string)
* Verify that we are actually talking to Google's servers when sending checkout commands 
* Generated XML will no longer include extra whitespace, to help comply with Google's XML schema
* Use BigDecimal instead of Float for all calculations due to weird rounding errors in Ruby's standard library
* Fixed deprecation warnings 

== 1.0.6.1 (2010-04-20)

* Fixed issue #14 - will not load on Ruby 1.9.x
* Added homepage URL to gem specification

== 1.0.6 (2010-04-01)

* Added beta Subscriptions API support.
* Fixed email-delivery XML generation bug.

== 1.0.5 (2008-9-17)

* Added Order Report API support.  Thanks to Daniel Pirone!
* Added unit tests for OrderReportCommand.  Added validation checks fro start_date, end_date, financial_state and fulfillment_state.

== 1.0.4 (2008-09-02)

* Fixed undefined email_delivery? method bug
* Added unit test to cover digital content creation

== 1.0.3 (2008-08-08)

* Added analytics_data support.  Thanks to Will Schwenk!
* Added create_address_filters_allowed_area and create_address_filters_excluded_area method to the MerchantCalculatedShipping class.
* Created two aliases create_shipping_restrictions_allowed_area and create_shipping_restrictions_excluded_area to facilitate deprecation of create_allowed_area and create_excluded_area.
* country_code in PostalArea class is now optional in the initialize method so it can be called properly in the create_area method in shared.rb.
* Added line-item shipping support.
* Refactor notifications to extract a base notification class that has the front_end, serial_number, google_order_number and timestamp.
* Added notification handshake support (i.e. serial-number) for notifciation-acknowledgement.
* Added digital delivery support.  Thanks to Andy Kim and Matthew Moore!
* Added Carrier Calculated Shipping support.
* Added unit tests - carrier_calculated_shipping_test.rb, digital_content_test.rb, *_items_*commands.rb
* Updated unit tests - item_test.rb, merchant_calculated_shipping_test.rb, notification_acknowledgement_test.rb
* Updated integration test - checkout_command_test.rb
* Added HTML Form API Signature support.  It is included in utils.rb.

== 1.0.2 (2007-08-18)

* Moved Address from notifications.rb to shared.rb
* Added AnonymousAddress class, now Address is extended from it
* Added merchant calcuations callback and result support
* Added unit tests for merchant calculation callback and result
* Added merchant calculations tag support in Shopping Cart XML
* Updated checkout_command_test.rb and checkout_command_xml_generator_test.rb after for testing merchant calculations tag support

== 1.0.1 (2007-08-16)

* Fixed text.to_s.gsub(/[^0-9]/, '') to (...text.to_f*100).to_i so dollar amount is now handled correctly

== 1.0.0 (2007-08-04)

* Added refund-amount-notification
* Added chargeback-amount-notification
* Added authorization-amount-notification
* Added refund command
* Added authorize command
* Added add-merchant-number command
* Added add-tracking-data command
* Added send-buyer-message command
* Added archive-order command
* Added unarchive-order command
* Refactored to_xml in the Command class
* Added CommandXmlGenerator base class
* Added ChargeOrderCommand, DeliverOrderCommand, CancelOrderCommand, ChargeAmountNotification and RiskInfoNotification (all contributed by Dan Dukeson).
* Attributes are not retrieved via XPaths with "tag/@attribute" any more but with REXML's Element#attributes. This should fix problems with Ruby 1.8.6.

== 0.1.0 (2007-05-12)
* Added property "shipping_taxed" to TaxRule (contributed by Dan Dukeson).
* Added XML generation for the tax tables in the shopping cart (contributed by Dan Dukeson).
* Added Area subclasses WorldArea and PostalArea (contributed by Dan Dukeson).
* Changed time zones in tests to UTC.
* Added Test::Unit::TestCase#assert_strings_equal to test_helper.rb.
* Added "rake test" (alias of "rake test:all")

test/test_helper.rb
# Project:   google4r
# File:      /test/test_helper.rb
# Author:    Manuel Holtgrewe <purestorm at ggnore dot net>
# Copyright: (c) 2007 by Manuel Holtgrewe
# License:   MIT License as follows:
#
# Permission is hereby granted, free of charge, to any person obtaining 
# a copy of this software and associated documentation files (the 
# "Software"), to deal in the Software without restriction, including 
# without limitation the rights to use, copy, modify, merge, publish, 
# distribute, sublicense, and/or sell copies of the Software, and to permit
# persons to whom the Software is furnished to do so, subject to the 
# following conditions:
#
# The above copyright notice and this permission notice shall be included 
# in all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
# OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF 
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. 
# IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY 
# CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, 
# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE
# OR THE USE OR OTHER DEALINGS IN THE SOFTWARE. 

# setup load path
$LOAD_PATH << File.join(File.dirname(__FILE__), '..')
$LOAD_PATH << File.join(File.dirname(__FILE__), '..', 'lib')

# require dependencies

require 'test/unit'

require 'rubygems'
require 'bundler'
Bundler.setup(:default, :test)

require 'mocha'
require 'nokogiri'
require 'pry'

class Array
  # Returns the "power set" for this Array. This means that an array with all
  # subsets of the array's elements will be returned.
  def power
    # the power set line is stolen from http://johncarrino.net/blog/2006/08/11/powerset-in-ruby/
    inject([[]]){|c,y|r=[];c.each{|i|r<<i;r<<i+[y]};r}
  end
end

# A helper class for the tests that require TaxTable objects.
class TestTaxTableFactory
  def effective_tax_tables_at(time)
    table = Google4R::Checkout::TaxTable.new(false)
    table.name = "Some Table"
    
    [ table ]
  end
end

class Test::Unit::TestCase
  def assert_nospace_equal(expected, actual)
    return assert_equal(expected.gsub(/\s/, ''), actual.gsub(/\s/, ''))
  end
  
  def command_selector(command)
    "#{Google4R::Checkout::CommandXmlGenerator::COMMAND_TO_TAG[command.class]}" +
    "[google-order-number='#{command.google_order_number}']"
  end
  
  def find_elements(selector, xml)
    Nokogiri.parse(xml).css(selector)
  end
  
  def assert_element_exists(selector, xml, msg=nil)
    found = find_elements(selector, xml)
    assert_not_equal 0, found.size, (msg || "Expected to find #{selector} in #{xml}")
  end
  
  def assert_no_element_exists(selector, xml, msg=nil)
    found = find_elements(selector, xml)
    assert_equal 0, found.size, (msg || "Expected to find #{selector} in #{xml}")
  end
  
  def assert_element_text_equals(text, selector, xml, msg=nil)
    found = find_elements(selector, xml)
    assert_equal 1, found.size, "Expected to find one #{selector} in #{xml} but found #{found.size}"
    assert_equal text, found.text, msg
  end
  
  def find_command_elements(selector, command)
    find_elements("#{command_selector(command)} #{selector}", command.to_xml)
  end
  
  def assert_command_element_text_equals(text, selector, command, msg=nil)
    assert_element_text_equals(text, "#{command_selector(command)} #{selector}", command.to_xml, msg)
  end
  
  def assert_command_element_exists(selector, command, msg=nil)
    assert_element_exists("#{command_selector(command)} #{selector}", command.to_xml, msg)
  end
  
  def assert_no_command_element_exists(selector, command, msg=nil)
    assert_no_element_exists("#{command_selector(command)} #{selector}", command.to_xml, msg)
  end
end


test/unit/item_test.rb
#--
# Project:   google_checkout4r 
# File:      test/unit/area_test.rb
# Author:    Manuel Holtgrewe <purestorm at ggnore dot net>
# Copyright: (c) 2007 by Manuel Holtgrewe
# License:   MIT License as follows:
#
# Permission is hereby granted, free of charge, to any person obtaining 
# a copy of this software and associated documentation files (the 
# "Software"), to deal in the Software without restriction, including 
# without limitation the rights to use, copy, modify, merge, publish, 
# distribute, sublicense, and/or sell copies of the Software, and to permit
# persons to whom the Software is furnished to do so, subject to the 
# following conditions:
#
# The above copyright notice and this permission notice shall be included 
# in all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
# OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF 
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. 
# IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY 
# CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, 
# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE
# OR THE USE OR OTHER DEALINGS IN THE SOFTWARE. 
#++

require File.expand_path(File.dirname(__FILE__)) + '/../test_helper'

require 'google4r/checkout'

require 'test/frontend_configuration'

# Test for the Item class.
class Google4R::Checkout::ItemTest < Test::Unit::TestCase
  include Google4R::Checkout

  def setup
    @frontend = Frontend.new(FRONTEND_CONFIGURATION)
    @frontend.tax_table_factory = TestTaxTableFactory.new
    
    @xml_str = %q{<?xml version="1.0" encoding="UTF-8" ?>
      <item>
        <item-name>MegaSound 2GB MP3 Player</item-name>
        <item-description>Portable MP3 player - stores 500 songs</item-description>
        <item-weight unit='LB' value='2.2'/>
        <unit-price currency="USD">178.00</unit-price>
        <quantity>1</quantity>
        <merchant-item-id>MGS2GBMP3</merchant-item-id>
        <merchant-private-item-data>
          <item-note>Text 1</item-note>
          <item-note>Text 2</item-note>
          <nested>
            <tags>value</tags>
          </nested>
        </merchant-private-item-data>
        <tax-table-selector>Some Table</tax-table-selector>
        <digital-content>
          <display-disposition>OPTIMISTIC</display-disposition>
          <email-delivery>true</email-delivery>
        </digital-content>
      </item>}
    
    @optional_tags = [ 'merchant-item-id', 'merchant-private-item-data', 'tax-table-selector' ]

    @command = @frontend.create_checkout_command
    @shopping_cart = @command.shopping_cart
    @item = @shopping_cart.create_item
    @digital_content = @item.digital_content
  end
  
  def test_item_behaves_correctly
    [ :shopping_cart,  :name, :name=, :description, :description=, :unit_price, :unit_price=,
      :quantity, :quantity=, :id, :id=, :private_data, :private_data=,
      :tax_table, :tax_table=, :digital_content, :weight, :weight=
    ].each do |symbol|
      assert_respond_to @item, symbol
    end
  end
  
  def test_item_gets_initialized_correctly
    assert_equal @shopping_cart, @item.shopping_cart
    assert_nil @item.name
    assert_nil @item.description
    assert_nil @item.unit_price
    assert_nil @item.quantity
    assert_nil @item.private_data
    assert_nil @item.id
    assert_nil @item.tax_table
    assert_nil @item.digital_content
  end
  
  def test_item_setters_work
    @item.name = "name"
    assert_equal "name", @item.name
    
    @item.description = "description"
    assert_equal "description", @item.description
    
    @item.unit_price = Money.new(100, "EUR")
    assert_equal Money.new(100, "EUR"), @item.unit_price
    
    @item.quantity = 10
    assert_equal 10, @item.quantity
    
    @item.id = "id"
    assert_equal "id", @item.id
    
    @item.private_data = Hash.new
    assert_equal Hash.new, @item.private_data
    
    @item.weight = Weight.new(2.2)
    assert_equal Weight, @item.weight.class
  end
  
  def test_set_tax_table_works
    table = @command.tax_tables.first
    @item.tax_table = table
    assert_equal table, @item.tax_table
  end
  
  def test_set_tax_table_raises_if_table_is_unknown_in_command
    assert_raises(RuntimeError) { @item.tax_table = TaxTable.new(false) }
  end
  
  def test_set_private_data_only_works_with_hashes
    assert_raises(RuntimeError) { @shopping_cart.private_data = 1 }
    assert_raises(RuntimeError) { @shopping_cart.private_data = nil }
    assert_raises(RuntimeError) { @shopping_cart.private_data = 'Foobar' }
    assert_raises(RuntimeError) { @shopping_cart.private_data = [] }
  end
  
  def test_item_price_must_be_money_instance
    assert_raises(RuntimeError) { @item.unit_price = nil }
    assert_raises(RuntimeError) { @item.unit_price = "String" }
    assert_raises(RuntimeError) { @item.unit_price = 10 }
  end

  def test_create_from_element_works
    @optional_tags.power.each do |optional_tag_names|
      xml_str = @xml_str

      optional_tag_names.each { |name| xml_str = xml_str.gsub(%r{<#{name}.*?</#{name}>}, '') }

      command = @frontend.create_checkout_command
      tax_table = TaxTable.new(false)
      tax_table.name = 'Some Table'
      command.tax_tables << tax_table
      item = Item.create_from_element(REXML::Document.new(xml_str).root, command.shopping_cart)
      
      assert_equal command.shopping_cart, item.shopping_cart
      
      assert_equal 'MegaSound 2GB MP3 Player', item.name
      assert_equal 'Portable MP3 player - stores 500 songs', item.description
      assert_equal Money.new(17800, 'USD'), item.unit_price
      assert_equal 1, item.quantity
      assert_equal 'MGS2GBMP3', item.id unless item.id.nil?
      assert_equal 2.2, item.weight.value
      assert_equal 'LB', item.weight.unit
      
      hash = 
        {
          'item-note' => [ 'Text 1', 'Text 2' ],
          'nested' => { 'tags' => 'value' }
        }
      assert_equal hash, item.private_data unless optional_tag_names.include?('merchant-private-item-data')

      assert_equal 'Some Table', item.tax_table.name unless optional_tag_names.include?('tax-table-selector')
    end
  end
end

lib/google4r/checkout/merchant_calculation.rb
#--
# Project:   google4r
# File:      lib/google4r/checkout/merchant_calculation.rb 
# Author:    Tony Chan <api dot htchan at gmail dot com>
# Copyright: (c) 2007 by Tony Chan
# License:   MIT License as follows:
#
# Permission is hereby granted, free of charge, to any person obtaining 
# a copy of this software and associated documentation files (the 
# "Software"), to deal in the Software without restriction, including 
# without limitation the rights to use, copy, modify, merge, publish, 
# distribute, sublicense, and/or sell copies of the Software, and to permit
# persons to whom the Software is furnished to do so, subject to the 
# following conditions:
#
# The above copyright notice and this permission notice shall be included 
# in all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
# OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF 
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. 
# IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY 
# CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, 
# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE
# OR THE USE OR OTHER DEALINGS IN THE SOFTWARE. 
#++
# This file contains the classes and modules that are used in the callback
# handling code.

require 'rexml/document'

module Google4R #:nodoc:
  module Checkout #:nodoc:
    # Thrown by Callback on unimplemented and unknown callbacks from Google.
    class UnknownCallbackType < Exception
    end
    
    # This class expects the message sent by Google. It parses the XMl document and returns
    # the appropriate callback. If the callback sent by Google is invalid then a 
    # UnknownCallbackType is raised that you should catch and then send a 404 to Google
    # to indicate that the callback handler has not been implemented yet.
    #
    # See http://code.google.com/apis/checkout/developer/index.html#merchant_calculations_api for
    # details.
    #
    # Note that you must protect the HTTPS request to the piece of code using a 
    # CallbackHandler by HTTP Auth Basic. If you are using Ruby On Rails then you can
    # use the great "simple_http_auth" plugin you can find here:
    # http://blog.codahale.com/2006/05/11/basic-http-authentication-with-rails-simple_http_auth/
    #
    # === Usage Example
    #
    # When you use a Rails controller to handle the calbacks by Google then your action to handle
    # the callbacks could use a CallbackHandler as follows:
    # 
    #  def google_checkout_api
    #    frontend = Google4R::Checkout::Frontend.new(FRONTEND_CONFIGURATION)
    #    frontend.tax_table_factory = TaxTableFactory.new
    #    handler = frontend.create_callback_handler
    #    
    #    begin
    #      callback = handler.handle(request.raw_post) # raw_post contains the XML
    #    rescue Google4R::Checkout::UnknownCallbackType => e
    #      # This can happen if Google adds new commands and Google4R has not been
    #      # upgraded yet. It is not fatal.
    #      render :text => 'ignoring unknown callback type', :status => 200
    #      return
    #    end
    #
    #    # ...
    #  end
    class CallbackHandler
      # The Frontend object that created this CallbackHandler
      attr_accessor :frontend
      
      # Create a new CallbackHandler and assign value of the parameter frontend to
      # the frontend attribute.
      def initialize(frontend)
        @frontend = frontend
      end
      
      # Parses the given xml_str and returns the appropriate *Callback class. At the
      # moment, only MerchantCalculationCallback objects can be returned.
      def handle(xml_str)
        root = REXML::Document.new(xml_str).root
        
        case root.name
        when 'merchant-calculation-callback' then
          MerchantCalculationCallback.create_from_element(root, frontend)
        else
          raise UnknownCallbackType, "Unknown callback type: #{root.name}"
        end
      end
    end
    
    # Google Checkout send a <merchant-calculation-callback> message to you when any of the
    # events occur:
    #   The customer signs in to Google Checkout.
    #   The customer enters a new shipping address on the Place Order page.
    #   The customer enters a coupon or gift certificate code.
    #
    # The message will be parsed into a MerchantCalculationCallback instance.
    #
    class MerchantCalculationCallback
      # The frontend this callback belongs to.
      attr_accessor :frontend

      # The order's shopping cart (ShoppingCart)
      attr_accessor :shopping_cart
      
      # The buyer's language
      attr_accessor :buyer_language
            
      # An array of AnonymousAddress objects of this callback.
      attr_reader :anonymous_addresses
         
      # This indicates whether the merchant needs to calculate taxes for the order.
      attr_accessor :tax
      
      # An array of shipping method names
      attr_reader :shipping_methods
      
      # An array of merchant codes
      attr_reader :merchant_code_strings
      
      # The tax tables for the items in the order notification.
      attr_reader :tax_tables
      
      # Sets the frontend attribute to the value of the frontend parameter.
      def initialize(frontend)
        @frontend = frontend
        @anonymous_addresses = Array.new
        @shipping_methods = Array.new
        @merchant_code_strings = Array.new
        if frontend.tax_table_factory
          @tax_tables = frontend.tax_table_factory.effective_tax_tables_at(Time.now)
        end
      end
      
      # Factory method to create a new MerchantCalculationCallback object from 
      # the REXML:Element object
      #
      # Raises NoMethodError and RuntimeError exceptions if the given element misses required
      # elements.
      #
      # You have to pass in the Frontend class this callback belongs to.
      def self.create_from_element(element, frontend)
        result = MerchantCalculationCallback.new(frontend)
        
        result.shopping_cart = ShoppingCart.create_from_element(element.elements['shopping-cart'], result)
        result.buyer_language = element.elements['buyer-language'].text
        element.elements.each('calculate/addresses/anonymous-address') do |address_element|
          result.anonymous_addresses << AnonymousAddress.create_from_element(address_element)
        end
        result.tax = element.elements['calculate/tax'].text
        element.elements.each('calculate/shipping/method') do |shipping_method_element|
          result.shipping_methods << shipping_method_element.attributes['name']
        end
        element.elements.each('calculate/merchant-code-strings/merchant-code-string') do |merchant_code_string_element|
          result.merchant_code_strings << merchant_code_string_element.attributes['code']
        end
        
        return result
      end
    end
    
    # This class represents a merchant-calculation-results XML
    # 
    # === Usage Sample
    #
    # results = MerchantCalculationResults.new
    # coupon_result = CouponResult.new(true, 'FirstVisitCoupon', Money.new(500, 'USD'), 'Congratulations! You saved $5.00 on your first visit!')
    # gift_certificate_result = GiftCertificateResult.new(true, 'GiftCert012345', Money.new(1000, 'USD'), 'You used your Gift Certificate!')
    #
    # results.create_merchant_calculation_result do |result|
    #   result.shipping_name = 'SuperShip'
    #   result.address_id = '739030698069958'
    #   result.shipping_rate = Money.new(703, 'USD')
    #   result.shippable = true
    #   result.total_tax = Money.new(1467, 'USD')
    #   result.create_merchant_code_result(@coupon_result)
    #   result.create_merchant_code_result(@gift_certificate_result)
    # end
    #
    # results.create_merchant_calculation_result do |result|
    #   result.shipping_name = 'UPS Ground'
    #   result.address_id = '739030698069958'
    #   result.shipping_rate = Money.new(556, 'USD')
    #   result.shippable = true
    #   result.total_tax = Money.new(1467, 'USD')
    #   result.create_merchant_code_result(@coupon_result)
    #   result.create_merchant_code_result(@gift_certificate_result)
    # end
    #
    # results.to_xml # To create the XML to return to Google
    #
    class MerchantCalculationResults
      # An array of merchant calcuation results
      attr_reader :merchant_calculation_results
      
      def initialize()
        @merchant_calculation_results = Array.new
      end
      
      # This method takes a MerchantCalculationResult object and add it to the
      # merchant_calculation_results array.  If the object is not provided, it will
      # instantiate one and add it to the array.  An optional code block can be
      # supplied to set the attributes of the new MerchantCalculationResult object.
      #
      # Raises RuntimeError exceptions if the given object is not of type
      # MerchantCalculationResult.
      def create_merchant_calculation_result(result=nil, &block)
        if result.nil?
          result = MerchantCalculationResult.new
          # Pass the newly generated item to the given block to set its attributes.
          yield(result) if block_given?
        else
          raise "Not a MerchantCalculationResult!" unless result.kind_of?(MerchantCalculationResult)
        end
        @merchant_calculation_results << result
      end
      
      def to_xml()
        return MerchantCalculationResultsXmlGenerator.new(self).generate()
      end
    end
    
    # The class represnts a merchant-calculation-result in the merchant-calculation-results XML
    class MerchantCalculationResult
      # The shipping name (string)
      attr_accessor :shipping_name
      
      # The address id (string)
      attr_accessor :address_id
      
      # The shipping rate (Money)
      attr_accessor :shipping_rate
      
      # Is it this applicable to this order (boolean)
      attr_accessor :shippable 
      
      # The total tax (Money)
      attr_accessor :total_tax
      
      # An array of merchant code results
      attr_reader :merchant_code_results
      
      def initialize(shipping_name='', address_id='', shipping_rate=nil, shippable=false, total_tax=nil)
        @shipping_name = shipping_name
        @address_id = address_id
        @shipping_rate = shipping_rate
        @shippable = shippable
        @total_tax = total_tax
        @merchant_code_results = Array.new
      end
      
      # This method takes either a CouponResult or GiftCertificateResult object and 
      # add it to the merchant_code_results array.  If the Class object of either
      # the two types is provided, it will create an instance from the Class object.
      # An optional code block can be supplied to set the attributes of the new 
      # object.
      #
      # Raises RuntimeError exceptions if there is no argument or a wrong class
      # type is provided.
      def create_merchant_code_result(result=nil, &block)
        if !result.nil?
          if [ CouponResult, GiftCertificateResult ].include?(result) # is a Class object
            result = result.new
          else
            raise "Invalid Merchant Code Result class type: #{result.class}!" unless 
              (result.kind_of?(CouponResult) || result.kind_of?(GiftCertificateResult))
          end
        else
          raise "You must either provide a MerchantCodeResult Class type or a CoupleResult or GiftCertificateResult instance."
        end
        @merchant_code_results << result

        # Pass the newly generated item to the given block to set its attributes.
        yield(result) if block_given?
        
      end
    end
    
    # Base class for merchant code (coupon and gift certificate)
    class CodeResult
      # Is this valid (boolean)
      attr_accessor :valid
      
      # The code (string)
      attr_accessor :code
      
      # The calculated amount (Money)
      attr_accessor :calculated_amount
      
      # The message (string)
      attr_accessor :message
      
      def initialize()
        raise "Do not use the abstract class Google::Checkout::CodeReslt"
      end
    end
    
    # This class represents a coupon-result in the merchant-calculation-results XML
    class CouponResult < CodeResult
      def initialize(valid=false, code='', calculated_amount=nil, message='')
        @valid = valid
        @code = code
        @calculated_amount = calculated_amount
        @message = message
      end
    end

    # This class represents a gift-certificate-result in the merchant-calculation-results XML
    class GiftCertificateResult < CodeResult
      def initialize(valid=false, code='', calculated_amount=nil, message='')
        @valid = valid
        @code = code
        @calculated_amount = calculated_amount
        @message = message
      end
    end
  end
end
